### ISSUE TITLE
[css-selectors] add numeric comparisons on attribute values

### ISSUE THREAD

**Issue by liamquin** [w3c/csswg-drafts#5890]
Can we have numeric comparisons?

e.g.
[data-balance<0] { color: red; } (common in accounting/finance, along with putting ( ) around the value)
[data-eyes>7.9] { color: black; background-colour: yellow } /* spider! */

with < <= >  >=

Noting that testing for equality is string equality so testing 06=6 fails, an alternative design might be to calc() to fetch an attribute value, and to be allowed in selectors, but that's more complex.

Maybe a new section inhttps://drafts.csswg.org/selectors/#attribute-selectors for Numeric Attribute Values?

**Comment by Loirooriol** [comment]
There is already #354, which has various proposals like this. But not sure if it's better to split them among multiple subissues, like this one.

**Comment by liamquin** [comment]
Thanks @Loirooriol - i don't know why i didn't find 354 when i was searching. On the other hand maybe considering numeric  values by themselves might make them easier to get  adopted.

**Comment by bramus** [comment]
_(Cross-post from #354)_

To numerically compare attribute values today, authors can work around this by:

1. Injecting all the values that the item is not equal to, using `data-*` attributes.
2. Using the CSS Attribute “Value Contains” Selector (e.g. `[attribute~=value]`) on those `data-*` attributes.

I’ve detailed this nasty and cumbersome hack back in the day over at https://www.bram.us/2016/10/13/css-attribute-value-less-than-greater-than-equals-selectors/.

**Comment by romainmenke** [comment]
Can a flag like `i` help?

- `[foo="bar" i]` matches cases insensitively
- `[foo="1e1" n]` switch to numeric comparison. (matches `<div foo="10">`, `<div foo="010">` `<div foo="10.00">`)

**Comment by Loirooriol** [comment]
It's not much clear how `n` would work with `^=`, `$=`, etc.
And it could cover some usecases, but it doesn't address inequalities < <= > >=

**Comment by tabatkins** [comment]
> It's not much clear how n would work with ^=, $=, etc.

I assume it wouldn't. Using it would be a syntax error.

> And it could cover some usecases, but it doesn't address inequalities < <= > >=

Right, inequalities are vital for this use-case, so this wouldn't really help.

A flag doesn't really add anything anyway - attribute selectors currently require the value to be an ident or a string, so numbers are invalid. Thus `[foo=5]` would already unambiguous as a numeric comparison rather than a string comparison.

**Comment by SebastianZ** [comment]
So far, I only saw positive feedback on this (here and in #354) and the proposed syntax seems clear. So let's add this to the agenda.

Proposal: Add the numeric comparison operators `<`, `<=`, `=`, `>=`, and `>` to attribute selectors.

Sebastian

**Comment by SebastianZ** [comment]
Note that I've removed the https://github.com/w3c/csswg-drafts/labels/selectors-4 label, as we're trying prepare Selectors 4 for CR status (see #12799). Therefore, this should rather be added to Selectors 5.

Sebastian

**Comment by Loirooriol** [comment]
Would this only work if the attribute value can be parsed as a number?

I think it would also be useful to e.g. match `width="10px"` with `[width > 5px]`.

Of course that raises the question of what to do with `em` units and such, I guess they could be based on the initial value like in https://drafts.csswg.org/mediaqueries-5/#units

**Comment by tabatkins** [comment]
Yes, I agree being able to use units would be nice. It's essentially identical from a parsing perspective, just allowing `<number> | <dimension> | <percentage>` rather than just `<number>`. Math functions too, then?

And yes, I expect that we'd resolve computed units like they were in MQs, and I guess %s would be left as-is (not resolving against anything), so we can use naive numeric comparison. (And using in a math function would simply not work here.)

**Comment by LeaVerou** [comment]
I think the semantics of `[foo > 5]` are clear without an additional opt-in. Adding support for dimensions adds a lot of complexity it for what seems like little benefit to me, and encourages problematic patterns. The vast majority of use cases I've encountered are around plain numbers. For dimensions, there's always the hack outlined in #354 so it's not like complex things aren't already possible. So let's focus on making simple things easy.

**Comment by romainmenke** [comment]
> Adding support for dimensions adds a lot of complexity it for what seems like little benefit to me, and encourages problematic patterns.

Can you specify?

@tabatkins said:

> attribute selectors currently require the value to be an ident or a string, so numbers are invalid. Thus `[foo=5]` would already unambiguous as a numeric comparison rather than a string comparison.

The same is true for dimension and function tokens, those are also invalid today.
So I am unsure where the extra complexity is that you see?

**Comment by Crissov** [comment]
> A flag doesn't really add anything anyway - attribute selectors currently require the value to be an ident or a string, so numbers are invalid. Thus `[foo=5]` would already unambiguous as a numeric comparison rather than a string comparison.

Huh, I actually assumed that despite what the spec says about [tokenization](https://drafts.csswg.org/css-syntax-3/#tokenization), current browsers were lenient with [attribute value selectors](https://drafts.csswg.org/selectors/#attribute-representation) and authors would be relying on unquoted values working like string tokens, because quotation marks are often optional in HTML. A [quick test](https://codepen.io/Crissov-the-bold/pen/ByKwJVm) reveals that @tabatkins is not only correct in theory but in practice, too (at least unless the CSS token starts with a letter). That’s great. 

Numeric comparison should therefore just work, i.e. run through the usual syntax processing steps for the host language and CSS first, so `+.0` may equal `0e-00` for instance.

For any more complex value type, there would still need to be some kind of method flag or type hint. 
For some types that exist in CSS syntax, like dimensions and percentages, the use of a respective token on the right-hand side of the comparison should suffice. The left-hand side would simply need to match, or at least be compatible (like percentages and numbers). 
For others, like colors or – not available in CSS – datetimes, this fails, because they cannot always be distinguished from other types, strings or numbers/equations in particular.  
(PS: For tuples like colors or coordinates and for non-numeric types, e.g. URLs, comparison with `<` and `>` is also not intuitively applicable, although one often _could_ specify some concept of order or magnitude for them.)

**Comment by tabatkins** [comment]
@romainmenke 
> The same is true for dimension and function tokens, those are also invalid today.
So I am unsure where the extra complexity is that you see?

I think you misunderstood my comment. I was saying that, since numeric values in attr selectors are today invalid, it's *perfectly fine* for us to allow `[foo = 5]` and have it be a *numeric* comparison rather than a string comparison, like `[foo="5"]` is today.

@Crissov 
> A [quick test](https://codepen.io/Crissov-the-bold/pen/ByKwJVm) reveals that @tabatkins is not only correct in theory but in practice, too (at least unless the CSS token starts with a letter).

Yeah, if the token starts with a letter it's an ident, which is valid (and we test string equality against its representation).

@LeaVerou 
> Adding support for dimensions adds a lot of complexity it for what seems like little benefit to me, and encourages problematic patterns.

I echo Romain's request for details here. What do you imagine is complicated here by using the MQ simplifications on both attr value and test value, and then requiring unit matches?

> For dimensions, there's always the hack outlined in https://github.com/w3c/csswg-drafts/issues/354

I reread the thread and can't figure out what hack you're referring to, could you elaborate?

**Comment by romainmenke** [comment]
I had to re-read my own comment several times before I realized the second part of my comment wasn't directed at you, @tabatkins, but rather I was quoting what you said as part of my reply to @LeaVerou 😅
